src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,365 @@
+                .expect("Failure to get right child index");
+            let left_hash = self.nodes.get(&left_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
+            let right_hash = self.nodes.get(&right_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
+            InnerNodeInfo {
+                value: **digest,
+                left: *left_hash,
+                right: *right_hash,
+            }
+        })
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds the nodes of the specified Merkle path to this [PartialMerkleTree]. The `index_value`
+    /// and `value` parameters specify the leaf node at which the path starts.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The depth of the specified node_index is greater than 64 or smaller than 1.
+    /// - The specified path is not consistent with other paths in the set (i.e., resolves to a
+    ///   different root).
+    pub fn add_path(
+        &mut self,
+        index_value: NodeIndex,
+        value: Word,
+        mut path: MerklePath,
+    ) -> Result<(), MerkleError> {
+        Self::check_depth(index_value.depth())?;
+        self.update_depth(index_value.depth());
+
+        // add node index to the leaves set
+        self.leaves.insert(index_value);
+        let sibling_node_index = Self::get_sibling_index(&index_value)?;
+        self.leaves.insert(sibling_node_index);
+
+        // add first two nodes to the nodes map
+        self.nodes.insert(index_value, value.into());
+        self.nodes.insert(sibling_node_index, path[0].into());
+
+        // update the current path
+        let parity = index_value.value() & 1;
+        path.insert(parity as usize, value);
+
+        // traverse to the root, updating the nodes
+        let mut index_value = index_value;
+        let root = Rpo256::merge(&[path[0].into(), path[1].into()]);
+        let root = path.iter().skip(2).copied().fold(root, |root, hash| {
+            index_value.move_up();
+            // insert calculated node to the nodes map
+            self.nodes.insert(index_value, root);
+
+            let sibling_node = Self::get_sibling_index_unchecked(&index_value);
+            // assume for now that all path nodes are leaves and add them to the leaves set
+            self.leaves.insert(sibling_node);
+
+            // insert node from Merkle path to the nodes map
+            self.nodes.insert(sibling_node, hash.into());
+
+            Rpo256::merge(&index_value.build_node(root, hash.into()))
+        });
+
+        let old_root = self.nodes.get(&Self::ROOT_INDEX).cloned().unwrap_or(Self::EMPTY_DIGEST);
+
+        // if the path set is empty (the root is all ZEROs), set the root to the root of the added
+        // path; otherwise, the root of the added path must be identical to the current root
+        if old_root == Self::EMPTY_DIGEST {
```

Comment:
Doesn't `self.root()` already kind of do this? So, we should be able to say something like:
```Rust
if self.root() == EMPTY_DIGEST {
```